File: src/tranquil/team_sync.py
from __future__ import annotations

import json
import urllib.request
from typing import Any

from .storage import Storage
from .util import iso_now


def push_sync(storage: Storage, endpoint: str, headers: dict[str, str] | None = None, timeout: float = 10.0) -> dict[str, Any]:
    data = storage.export_data()
    payload = {
        "schema": "tranquil.sync/v1",
        "exported_at": iso_now(),
        "data": data,
    }
    body = json.dumps(payload, sort_keys=True, ensure_ascii=False).encode("utf-8")
    request = urllib.request.Request(
        endpoint,
        data=body,
        headers={"Content-Type": "application/json", **(headers or {})},
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=timeout) as response:
        response_body = response.read().decode("utf-8")
        status = response.status
    return {
        "status": status,
        "events": len(data["events"]),
        "runs": len(data["runs"]),
        "fixtures": len(data["fixtures"]),
        "scores": len(data["scores"]),
        "response": response_body[:2000],
    }

File: rust/src/team_sync.rs
use serde_json::{Value, json};

use crate::http::post_json;
use crate::storage::Storage;
use crate::util::now_ts;

pub struct SyncResult {
    pub status: u16,
    pub events: usize,
    pub runs: usize,
    pub fixtures: usize,
    pub scores: usize,
}

pub fn push_sync(
    storage: &Storage,
    endpoint: &str,
    headers: &[(String, String)],
) -> Result<SyncResult, String> {
    let data = storage.export_data()?;
    let events = len(&data, "events");
    let runs = len(&data, "runs");
    let fixtures = len(&data, "fixtures");
    let scores = len(&data, "scores");
    let payload = json!({
        "schema": "tranquil.sync/v1",
        "exported_at": now_ts(),
        "data": data,
    });
    let response = post_json(
        endpoint,
        headers,
        serde_json::to_string(&payload).map_err(|err| err.to_string())?,
        10,
    )?;
    Ok(SyncResult {
        status: response.status,
        events,
        runs,
        fixtures,
        scores,
    })
}

fn len(value: &Value, key: &str) -> usize {
    value
        .get(key)
        .and_then(Value::as_array)
        .map(Vec::len)
        .unwrap_or(0)
}
